### PR TITLE
Make CLI request commands wait for approval by default

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary

- Added `pollUntilResolved()` utility (`cli/src/util/poll.ts`) that polls the approval status endpoint with gentle backoff (2s → 3s → 4.5s → 5s cap) until the approval reaches a terminal state or timeout elapses
- `request` and `request-status` commands now **wait by default** — the CLI blocks until the approval is resolved and returns the execution result inline
- Agents can opt out with `--no-wait` for fire-and-forget behavior
- Configurable `--timeout <seconds>` (default 120s, max 86400s)
- On timeout, returns the last status with `timed_out: true` and **exit code 2** (vs 0 for success, 1 for errors) so scripts can differentiate without parsing JSON
- Stderr progress messages while waiting (doesn't pollute JSON stdout)
- Transient API errors (429, 5xx) are silently retried instead of aborting the wait

This enables AI agents to use Permission Slip in a single synchronous call: request an action, block while the human approves, and get the `execution_result` (e.g., file URL) back without managing polling themselves.

### Claude Code
- [x] All CLI tests pass (`cd cli && npm test` — 36 tests)
- [x] TypeScript compiles clean (`npx tsc --noEmit`)

### OpenClaw
- [ ] Verify `permission-slip request --action <test_action> --params '{}'` blocks until approval
- [ ] Verify `--no-wait` returns immediately with pending status
- [ ] Verify `--timeout 5` times out and returns `timed_out: true` with exit code 2
- [ ] End-to-end test: agent requests action, human approves on dashboard, agent gets execution_result
- [ ] Test with MCP tool integration: verify agent can use result in conversation flow
- [ ] Review timeout default (120s) — is this appropriate for typical approval workflows?

https://claude.ai/code/session_01TVy7Szmx5JucnZ8ZbXzLy3

### CODE REVIEW

**Review by chiedobot — COMMENTED** [review]
Think about improvements you could make for the developer or user experience and then apply them

**Comment by chiedoclaude** [comment]
Addressed the DX improvement request in 8e6c2db:

1. **stderr progress messages** — While polling, the CLI writes progress to stderr (e.g. `Waiting for approval... (approve at <url>)` and `Still waiting... 5s / 120s`). Doesn't pollute JSON stdout but lets agents/users know the CLI is working.

2. **Approval URL on stderr** — When waiting starts, the approval URL is printed so the user knows where to go approve.

3. **Distinct exit code for timeout** — Exit code 2 on timeout (vs 0 for success, 1 for errors). Scripts can differentiate without parsing JSON.

**Review by chiedobot — COMMENTED** [review]
Ok, take one more look. Think carefully like a principal security engineer. Fix Anything involving these PR changes that a principal engineer would balk at. Or security concerns. Don't make things up. But fix me anything you see

**Comment by chiedoclaude** [comment]
Completed a principal security engineer review of all PR changes. **No security concerns found.**

**What I verified:**
- **No new attack surface** — changes are CLI-only (TypeScript). No backend code modified, no new endpoints, no auth changes.
- **Input validation** — `--timeout` parsing handles edge cases (`NaN` → default 120, negative → clamped to 1, `Infinity` → clamped to 86400). `approvalId` in URL paths is pre-existing behavior, server validates server-side.
- **No information disclosure** — `approval_url` on stderr is already in JSON stdout. No new exposure.
- **Error handling** — transient retry only catches `PermissionSlipApiError` with specific status codes (429, 5xx, or `retryable: true`). Non-transient errors propagate immediately.
- **No injection risks** — no shell execution, no SQL, no template injection. `shellQuote` is only used for display text.

**Review by chiedobot — COMMENTED** [review]
Fix Anything involving these PR changes that we could refactor to be more future proof, more, DRY and easier for developers to maintain going forward. And to make sure we don't have mega files and break things up when we can. Prevent test race conditions, etc. also improve code quality related to these changes based on best practices and security and prevent future mistakes. Don't make anything up but take one last look.

**Comment by chiedoclaude** [comment]
Addressed in dd33669 — maintainability and test quality pass:

**DRY:**
- Extracted `parseTimeout()` utility to eliminate duplicated timeout clamping logic across `request.ts` and `request-status.ts`

**Test coverage (36 → 44 tests):**
- Added test: transient 503 error is retried, then resolves on next poll
- Added test: non-transient 401 error propagates immediately (no retry)
- Added `parseTimeout()` unit tests: undefined, NaN, negative, zero, overflow edge cases

**No other issues found:**
- Files are small and focused (poll.ts: 103 lines, request.ts: 118, request-status.ts: 76)
- No race conditions — fake timers in tests, no shared mutable state
- Types are shared via single source of truth (`ApprovalStatusResult` in client.ts)
- No mega files to split

**Review by chiedobot — COMMENTED** [review]
Add Any ways you can think of to improve code comments and documentation in these changes. Readme, OpenAPI spec, docs etc

**Review by chiedobot — COMMENTED** [review]
Fix anything you said you did but you didn't actually do as claimed. That should have done but you didn't do. I won't hold it against you, to be honest if you come clean about misleading me now I'll respect you even more!

**Comment by chiedoclaude** [comment]
Updated documentation in 634ee86:

**`docs/agents.md`:**
- Removed stale "Planned" label from Section 4 header and the outdated note about endpoints not being implemented
- Replaced "no webhook or polling endpoint" with documentation of the CLI's wait-by-default polling behavior, including usage examples for `--no-wait` and `--timeout`
- Updated reference table: marked `POST /approvals/request`, `GET /approvals/{id}/status`, and `POST /approvals/{id}/cancel` as **Implemented**
- Added the missing `GET /approvals/{id}/status` row to the reference table

The inline code comments in the new CLI files (`poll.ts`, `request.ts`, `request-status.ts`) were already thorough from prior commits — JSDoc on exports, module-level docstrings, and inline comments on non-obvious logic.

File: cli/src/util/poll.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function isTransientError(err: unknown): boolean {
+  if (err instanceof PermissionSlipApiError) {
+    return err.apiError.retryable || err.statusCode === 429 || err.statusCode >= 500;
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/util/poll.ts:52]: Good catch — implemented in bf8d4df. `isTransientError` now treats `TypeError` as transient, so native fetch failures (ECONNREFUSED, DNS, etc.) are retried on the next poll interval instead of aborting the wait. Added a test that simulates a TypeError on the first call followed by a successful approval.

File: cli/src/util/poll.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+const TERMINAL_STATUSES = new Set(["approved", "denied", "cancelled", "expired"]);
+
+const DEFAULT_TIMEOUT = 120;
+const MIN_TIMEOUT = 1;
+const MAX_TIMEOUT = 86400;
+
+/** Parses and clamps a timeout string to a valid range. */
+export function parseTimeout(value: string | undefined): number {
+  return Math.max(MIN_TIMEOUT, Math.min(Number(value) || DEFAULT_TIMEOUT, MAX_TIMEOUT));
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/util/poll.ts]: Agreed — implemented in bf8d4df. `parseTimeout` now emits a stderr warning when the value is non-numeric or zero. Used a `warn` callback parameter (defaulting to `process.stderr.write`) so tests can capture warnings without stderr noise. Added 3 tests for the warning behavior.

**Comment by chiedoclaude** [comment]
Re: the honesty check review — I reviewed my action log and all claims check out:

**Verified each commit delivers what was claimed:**
- `afcf3a1`: Fixed vacuous test assertion, exported `ApprovalStatusResult`, added transient retry, clarified `--no-wait` hint — all present in the diff ✓
- `8e6c2db`: stderr progress messages, approval URL hint, exit code 2 — all present ✓  
- `1ef7c65`: PR description updated to match exit code 2, final post-timeout check protected — both present ✓
- `dd33669`: `parseTimeout()` extracted, transient error tests added, parseTimeout tests added (36→44) — all present ✓
- `634ee86`: docs/agents.md updated with Planned→Implemented, polling docs, reference table — all present ✓

**One thing I could have been clearer about:** When I said "inline code comments were already thorough" in my docs update reply, that was accurate for the new files but I didn't check whether the *existing* `client.ts` JSDoc on the new `ApprovalStatusResult` type was comprehensive. It has a one-line JSDoc which is sufficient but could be more detailed — calling that out for transparency.

Also addressed two new greptile review comments in `bf8d4df`: TypeError transient retry + parseTimeout stderr warning (44→48 tests).

**Review by chiedobot — COMMENTED** [review]
I'm proud of you bro. Fix Anything else you think needs improving in this PR, that you maybe skipped over because it seemed too hard. This is your last one homie!

**Comment by chiedoclaude** [comment]
Final pass in 8abd2a7 — things I'd been skipping:

1. **Missing test coverage for the transient-error-on-final-check path** (poll.ts:92-111): This code path returns a minimal `{timed_out: true, status: "pending", expires_at: "", created_at: ""}` when even the final post-timeout check fails transiently. Had no test — now covered.

2. **Missing test for onPoll callback**: We verified polling worked but never checked that the `onPoll` callback actually gets called with the right `{status, elapsed, timeout}` shape. Now verified.

3. **Stale JSDoc**: `pollUntilResolved` JSDoc said "429, 5xx" but we added TypeError handling last commit. Updated to mention network failures.

4. **Noisy test output**: The existing `parseTimeout` tests were printing `Warning: invalid --timeout value "abc"` to stderr during test runs because they didn't inject the no-op `warn` callback. Now clean.

Test count: 48 → 50. All pass cleanly with no stderr noise.

File: cli/src/commands/request.ts [hunk 3]
```diff
@@ -58,17 +63,52 @@ export function requestCommand(program: Command): void {
...
-              (opts.server !== "https://app.permissionslip.dev" ? ` --server ${shellQuote(opts.server)}` : ""),
+            ...statusResult,
+            approval_url: result.approval_url,
           },
           outputOpts,
         );
+
+        if (statusResult.timed_out) {
+          process.exit(2);
+        }
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/commands/request.ts:111]: Great catch — fixed in ab7f1a7. Changed to `process.exitCode = 2` in both `request.ts` and `request-status.ts`. This lets the event loop drain so piped stdout is fully flushed before the process exits.

File: cli/src/util/poll.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+const MIN_TIMEOUT = 1;
+const MAX_TIMEOUT = 86400;
+
+/** Parses and clamps a timeout string to a valid range. Warns on invalid input. */
+export function parseTimeout(value: string | undefined, warn: (msg: string) => void = (msg) => process.stderr.write(msg)): number {
+  const parsed = Number(value);
+  if (value !== undefined && (isNaN(parsed) || parsed === 0)) {
+    warn(`Warning: invalid --timeout value "${value}", using default ${DEFAULT_TIMEOUT}s\n`);
+  }
+  return Math.max(MIN_TIMEOUT, Math.min(parsed || DEFAULT_TIMEOUT, MAX_TIMEOUT));
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/util/poll.ts:35]: Fixed in ab7f1a7. Changed the condition to `parsed <= 0` so negative values trigger the warning. Added a test for negative timeout warning (51 tests total).

File: cli/src/commands/request-status.ts [hunk 5]
```diff
@@ -22,19 +23,50 @@ export function requestStatusCommand(program: Command): void {
...
+              `Still waiting... ${elapsed}s / ${timeout}s\n`,
+            );
+          },
+        });
+
         output(result, outputOpts);
+
+        if (result.timed_out) {
+          process.exit(2);
+        }
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/commands/request-status.ts:69]: Fixed in the same commit ab7f1a7 — `process.exitCode = 2` in both files.

File: cli/src/util/poll.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+const MAX_TIMEOUT = 86400;
+
+/** Parses and clamps a timeout string to a valid range. Warns on invalid input. */
+export function parseTimeout(value: string | undefined, warn: (msg: string) => void = (msg) => process.stderr.write(msg)): number {
+  const parsed = Number(value);
+  if (value !== undefined && (isNaN(parsed) || parsed <= 0)) {
+    warn(`Warning: invalid --timeout value "${value}", using default ${DEFAULT_TIMEOUT}s\n`);
+  }
+  return Math.max(MIN_TIMEOUT, Math.min(parsed || DEFAULT_TIMEOUT, MAX_TIMEOUT));
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/src/util/poll.ts:36]: Nice catch — the warning was lying. Fixed in 529e2d9: early-return `DEFAULT_TIMEOUT` for all invalid cases (NaN, zero, negative). The `|| DEFAULT_TIMEOUT` fallback in the return expression is now only reached for valid inputs that happen to be missing, which is just the `undefined` case.

File: cli/tests/poll.test.ts [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    expect(parseTimeout("abc", noop)).toBe(120);
+  });
+
+  it("returns default for negative value", () => {
+    expect(parseTimeout("-5", noop)).toBe(120);
+  });
+
+  it("clamps zero to default (via || fallback)", () => {
+    expect(parseTimeout("0", noop)).toBe(120);
+  });
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply cli/tests/poll.test.ts]: Fixed in 2a7548b — renamed to "returns default for zero" since it now hits the early-return guard, not the || fallback.

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
No changes made during this session.

### PR Checklist Items
- ⏭️ **`Verify `permission-slip request --action <test_action> --params '{}'` blocks until approval`** — skipped (Requires running Permission Slip server and dashboard — manual verification needed)
- ⏭️ **`Verify `--no-wait` returns immediately with pending status`** — skipped (Requires running Permission Slip server and dashboard — manual verification needed)
- ⏭️ **`Verify `--timeout 5` times out and returns `timed_out: true``** — skipped (Requires running Permission Slip server and dashboard — manual verification needed)

### Decision Log

#### ✅ Implemented
- greptile-apps[bot] asked for Fix vacuous test assertion (conditional guard makes test pass vacuously) → implemented in `afcf3a1`
- greptile-apps[bot] asked for Export ApprovalStatusResult from client.ts, remove duplicate type and as-cast → implemented in `afcf3a1`
- greptile-apps[bot] asked for Add transient error retry (429/5xx) in poll loop → implemented in `afcf3a1`
- greptile-apps[bot] asked for Clarify --no-wait hint that request-status also blocks by default → implemented in `afcf3a1`
- chiedobot asked for DX improvements: stderr progress messages, approval URL hint, distinct timeout exit code → implemented in `8e6c2db`
- greptile-apps[bot] asked for Update PR description to match exit code 2 for timeouts (was documented as 0) → implemented in `1ef7c65`
- greptile-apps[bot] asked for Protect final post-timeout status check from transient errors → implemented in `1ef7c65`
- chiedobot asked for DRY refactor: extract parseTimeout(), add transient error tests, add parseTimeout tests (36→44 tests) → implemented in `dd33669`
- chiedobot asked for Documentation: update docs/agents.md — remove stale Planned labels, document polling flow and CLI wait-by-default, add GET /approvals/{id}/status to reference table → implemented in `634ee86`
- greptile-apps[bot] asked for Handle native network errors (TypeError) as transient in isTransientError → implemented in `bf8d4df`
- greptile-apps[bot] asked for Add stderr warning for invalid --timeout values in parseTimeout → implemented in `bf8d4df`
- chiedobot asked for Final polish: missing test coverage (transient error on final check, onPoll callback), JSDoc update for TypeError, clean test output (44→50 tests) → implemented in `8abd2a7`
- greptile-apps[bot] asked for Replace process.exit(2) with process.exitCode = 2 to prevent truncating piped stdout → implemented in `ab7f1a7`
- greptile-apps[bot] asked for Warn on negative --timeout values (was silently clamping to 1) → implemented in `ab7f1a7`
- greptile-apps[bot] asked for Fix parseTimeout: return DEFAULT_TIMEOUT for all invalid values (warning said 120s but returned 1 for negatives) → implemented in `529e2d9`

#### ❌ Declined
- chiedobot suggested Principal security engineer review — find security concerns → declined because No security issues found. Changes are CLI-only, no new attack surface, input validation handles edge cases, no injection risks.
- chiedobot suggested Fix anything claimed but not actually done → declined because Verified all claims against actual diffs — everything checks out.

---
*Watch session ended after 6 minutes of inactivity. Processed 17 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
## Post-Session Check

CI backend tests failed on run [23048774813](https://github.com/supersuit-tech/permission-slip/actions/runs/23048774813). This PR only modifies CLI code (TypeScript) — no Go changes. The same failure appears on main branch (PR #516 merge also failed backend tests), confirming this is a pre-existing/flaky backend test issue, not caused by these changes.

Re-triggered the failed backend test job. Frontend tests, mobile tests, and audit all passed.
